### PR TITLE
OpenTelemetry: add AttributesCache to reduce memory allocations

### PR DESCRIPTION
**Pull request by brunokoeferli**

### CODE REVIEW

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * </pre>
+ * </p>
+ * <p>
+ * <b>Important:</b> Do only use a cache for low cardinality attributes. Otherwise the memory usage of the cache will
+ * cause more issues than it solves.
+ * </p>
+ *
+ * @see Attributes
+ */
+public class AttributesCache<KEY1, KEY2> {
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java:43]: Add unit test covering basic cache features of this cache.

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  public AttributesCache(int initialCapacityDimension1, int initialCapacityDimension2, BiFunction<KEY1, KEY2, Attributes> createAttributesFunction) {
+    m_cache = new ConcurrentHashMap<>(initialCapacityDimension1);
+    m_newMapFunction = k -> new ConcurrentHashMap<>(initialCapacityDimension2);
+    m_createAttributesFunction = createAttributesFunction;
+  }
+
+  public Attributes getAndPut(KEY1 key1, KEY2 key2) {
+    Map<KEY2, Attributes> attributesMap = m_cache.computeIfAbsent(key1, m_newMapFunction);
+    // primitive "computeIfAbsent" implementation to avoid memory allocation here (e.g. to create a mapping function for each call)
+    Attributes attributes = attributesMap.get(key2);
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java:69]: this is not 100% thread safe, two caller could get a different attributes instance if both encounter (attributes == null) and create a new attributes instance.
If this should be 100% thread safe you could use double checking -> first null check without synchronization, then enter a synchronized block and re-check

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    if (attributes == null) {
+      attributes = m_createAttributesFunction.apply(key1, key2);
+      attributesMap.put(key2, attributes);
+    }
+    return attributes;
+  }
+
+  public void put(KEY1 key1, KEY2 key2) {
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java:80]: Javadoc indicating difference to getAndPut, e.g. "use this method to pre-populate cache for given key1/2"

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  private final Function<KEY1, Map<KEY2, Attributes>> m_newMapFunction;
+  private final BiFunction<KEY1, KEY2, Attributes> m_createAttributesFunction;
+
+  public AttributesCache(int initialCapacityDimension1, int initialCapacityDimension2, BiFunction<KEY1, KEY2, Attributes> createAttributesFunction) {
+    m_cache = new ConcurrentHashMap<>(initialCapacityDimension1);
+    m_newMapFunction = k -> new ConcurrentHashMap<>(initialCapacityDimension2);
+    m_createAttributesFunction = createAttributesFunction;
+  }
+
+  public Attributes getAndPut(KEY1 key1, KEY2 key2) {
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java]: Javadoc: e.g. "return caches Attributes instance, if absent, a new instance is created and added to the cache"

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java (hunk 1 above)
Comment (brunokoeferli) [reply org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java:43]: done

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java (hunk 2 above)
Comment (brunokoeferli) [reply org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java:69]: this is by intension. i prefer creating the attributes instance several times and do without thread synchronization. i've noted this in the javadoc.

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java (hunk 3 above)
Comment (brunokoeferli) [reply org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java:80]: done

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java (hunk 4 above)
Comment (brunokoeferli) [reply org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java]: done

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  private final BiFunction<KEY1, KEY2, Attributes> m_createAttributesFunction;
+
+  protected AttributesCache(int initialCapacityDimension1, int initialCapacityDimension2, BiFunction<KEY1, KEY2, Attributes> createAttributesFunction) {
+    m_cache = new ConcurrentHashMap<>(initialCapacityDimension1);
+    m_newMapFunction = k -> new ConcurrentHashMap<>(initialCapacityDimension2);
+    m_createAttributesFunction = createAttributesFunction;
+  }
+
+  /**
+   * Return caches {@link Attributes} instance, if absent, a new instance is created and added to the cache.
```
Comment (paolobazzi) [inline org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/opentelemetry/AttributesCache.java]: "cached"

**Review by paolobazzi — APPROVED** [review]
Looks good!

**Pull request merged**